### PR TITLE
[TEVA-3117] Replace Postgres instance

### PR DESCRIPTION
**Pull request by felixtheflex**
## Jira ticket URL

- Just add the ticket number to the end:

https://dfedigital.atlassian.net/browse/TEVA-

## Changes in this PR:

- Is there anything specific you want feedback on?

## Next steps:

- [ ] Terraform deployment required?

- [ ] New development configuration to be shared?

### CODE REVIEW

File: .github/workflows/recreate-qa-database.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      shell: bash
+      run: |
+        gem install aws-sdk-ssm --no-document
+        bin/fetch_config.rb -s aws-ssm-parameter-path:/teaching-vacancies/${{ env.PARAMETER_STORE_ENVIRONMENT }}/app -d quiet \
+          && echo Data in "/teaching-vacancies/${{ env.PARAMETER_STORE_ENVIRONMENT }}" looks valid
+
+    - name: Deploy to environment
+      if: ${{ github.event.inputs.environment }} != "production"
+      run: |
+        make qa ci terraform-app-database-replace CONFIRM_REPLACE=yes
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:80]: Shouldn't we use env.environment instead of qa?

File: .github/workflows/recreate-qa-database.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      uses: hashicorp/setup-terraform@v1
+      with:
+        terraform_version: 0.15.5
+
+    - name: Set environment variables for non-review environments
+      run: |
+        PARAMETER_STORE_ENVIRONMENT=${{ github.event.inputs.environment }}
+        DEPLOY_ENV=${{ github.event.inputs.environment }}
+        echo "PARAMETER_STORE_ENVIRONMENT=${PARAMETER_STORE_ENVIRONMENT}" >> $GITHUB_ENV
+        echo "DEPLOY_ENV=${DEPLOY_ENV}" >> $GITHUB_ENV
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:54]: Is DEPLOY_ENV used anywhere?

File: .github/workflows/recreate-qa-database.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        CONFIRM_REPLACE=yes
+        echo "CONFIRM_REPLACE=${CONFIRM_REPLACE}" >> $GITHUB_ENV
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:56]: Is env.CONFIRM_REPLACE used anywhere?

File: Makefile [hunk 4]
```diff
@@ -117,6 +117,7 @@ ci:	## Run in automation environment
 
 .PHONY: terraform-app-init
 terraform-app-init:
+		$(if $(tag), , $(eval export tag=master))
```
Comment (saliceti) [inline Makefile:120]: Can this be moved to a separate commit to emphasize the new behaviour? And this would be helpful to document it.

File: Makefile [hunk 5]
```diff
@@ -133,6 +134,10 @@ terraform-app-destroy: terraform-app-init ## make qa destroy passcode=MyPasscode
 	$(if $(CONFIRM_DESTROY), , $(error Can only run with CONFIRM_DESTROY))
 	cd terraform/app && terraform destroy -var-file ../workspace-variables/${var_file}.tfvars
 
+terraform-app-database-replace: terraform-app-init check-docker-tag
+	$(if $(CONFIRM_REPLACE), , $(error Can only run with CONFIRM_REPLACE))
+	cd terraform/app && terraform apply -replace="module.paas.cloudfoundry_service_instance.postgres_instance" -replace="module.paas.cloudfoundry_app.web_app" -replace="module.paas.cloudfoundry_app.worker_app" -var-file ../workspace-variables/${var_file}.tfvars -auto-approve
```
Comment (saliceti) [inline Makefile:139]: Can we use line breaks to improve readability?

File: .github/workflows/recreate-qa-database.yml (hunk 2 above)
Comment (felixtheflex) [reply .github/workflows/recreate-qa-database.yml:54]: It would be if Idecided to say:-
make ${{env.DEPLOY_ENV }} ci terraform-app-database-replace CONFIRM_REPLACE=yes.............. and remove `qa`

Comment (felixtheflex) [reply .github/workflows/recreate-qa-database.yml:54]: or remove it and use `make ${{ github.event.inputs.environment }} ......`

File: .github/workflows/recreate-qa-database.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    - name: Pin Terraform version
+      uses: hashicorp/setup-terraform@v1
+      with:
+        terraform_version: 0.15.5
+
+    - name: Set environment variables for non-review environments
+      run: |
+        PARAMETER_STORE_ENVIRONMENT=${{ github.event.inputs.environment }}
+        DEPLOY_ENV=${{ github.event.inputs.environment }}
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:52]: This line can be removed I guess

File: .github/workflows/recreate-qa-database.yml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      environment:
+        description: Environment
+        default: qa
+        required: true
+
+env:
+ DOCKER_REPOSITORY: ghcr.io/dfe-digital/teaching-vacancies
+
+jobs:
+  turnstyle:
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:15]: Do we need a separate job? Turnstyle could be a step in the main job

File: .github/workflows/recreate-qa-database.yml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    steps:
+    - uses: DFE-Digital/github-actions/turnstyle@master
+      name: Check workflow concurrency
+      with:
+        initial-wait-seconds: 12
+        poll-interval-seconds: 20
+        same-branch-only: true
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  deploy-app:
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:27]: The job name should be updated

File: Makefile [hunk 9]
```diff
@@ -133,6 +134,12 @@ terraform-app-destroy: terraform-app-init ## make qa destroy passcode=MyPasscode
 	$(if $(CONFIRM_DESTROY), , $(error Can only run with CONFIRM_DESTROY))
 	cd terraform/app && terraform destroy -var-file ../workspace-variables/${var_file}.tfvars
 
+terraform-app-database-replace: terraform-app-init check-docker-tag
+	$(if $(CONFIRM_REPLACE), , $(error Can only run with CONFIRM_REPLACE)) \
+	cd terraform/app && terraform apply -replace="module.paas.cloudfoundry_service_instance.postgres_instance" \
```
Comment (saliceti) [inline Makefile:139]: Can we indent these lines to show they are part of the main command?

File: .github/workflows/recreate-qa-database.yml (hunk 7 above)
Comment (saliceti) [reply .github/workflows/recreate-qa-database.yml:15]: This could be changed similarly here: https://github.com/DFE-Digital/teaching-vacancies/blob/master/.github/workflows/deploy_app.yml#L20

File: .github/workflows/recreate-qa-database.yml [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        description: Environment
+        default: qa
+        required: true
+
+env:
+ DOCKER_REPOSITORY: ghcr.io/dfe-digital/teaching-vacancies
+
+jobs:
+
+  deploy-app:
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:16]: Can you update the job title and name below?

File: .github/workflows/recreate-qa-database.yml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    name: Deploy master image tag to ${{ github.event.inputs.environment }}
+    runs-on: ubuntu-20.04
+    steps:
+
+    - uses: DFE-Digital/github-actions/turnstyle@master
+      timeout-minutes: 20
```
Comment (saliceti) [inline .github/workflows/recreate-qa-database.yml:22]: As discussed, can you replacing the job property timeout-minutes with abort-after-seconds step property

Comment (saliceti) [reply .github/workflows/recreate-qa-database.yml:22]: And change also: https://github.com/DFE-Digital/teaching-vacancies/blob/master/.github/workflows/deploy_app.yml#L20

**Review by saliceti — APPROVED** [review]
